zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1233 @@
+    fixed::FixedMulDiv,
+};
+
+/// Gets the maximum number of nodes allowed in the liquidity tree as a function of its depth.
+/// Saturates at `u32::MAX`, but will warn about this in DEBUG.
+///
+/// # Generics
+///
+/// - `D`: A getter for the depth of the tree.
+pub struct LiquidityTreeMaxNodes<D>(PhantomData<D>);
+
+impl<D> Get<u32> for LiquidityTreeMaxNodes<D>
+where
+    D: Get<u32>,
+{
+    fn get() -> u32 {
+        debug_assert!(D::get() < 31, "LiquidityTreeMaxNodes::get(): Integer overflow");
+        2u32.saturating_pow(D::get() + 1).saturating_sub(1)
+    }
+}
+
+/// A segment tree used to track balances of liquidity shares which allows `O(log(n))` distribution
+/// of fees.
+///
+/// Each liquidity provider owns exactly one node of the tree which records their stake and fees.
+/// When a liquidity provider leaves the tree, the node is not removed from the tree, but marked as
+/// _abandoned_ instead. Abandoned nodes are reassigned when new LPs enter the tree. Nodes are added
+/// to the leaves of the tree only if there are no abandoned nodes to reassign.
+///
+/// Fees are lazily propagated down the tree. This allows fees to be deposited to the tree in `O(1)`
+/// (fees deposited at the root and later propagated down). If a particular node requires to know
+/// what its fees are, propagating fees to this node takes `O(depth)` operations (or, equivalently,
+/// `O(log_2(node_count))`).
+///
+/// # Generics
+///
+/// - `T`: The pallet configuration.
+/// - `U`: A getter for the maximum depth of the tree. Using a depth larger than `31` will result in
+///   undefined behavior.
+#[derive(
+    CloneNoBound, Decode, Encode, Eq, MaxEncodedLen, PartialEqNoBound, RuntimeDebugNoBound, TypeInfo,
+)]
+#[scale_info(skip_type_params(T, U))]
+pub struct LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    /// A vector which holds the nodes of the tree. The nodes are ordered by depth (the root is the
+    /// first element of `nodes`) and from left to right. For example, the right-most grandchild of
+    /// the root is at index `6`.
+    pub(crate) nodes: BoundedVec<Node<T>, LiquidityTreeMaxNodes<U>>,
+    /// Maps an account to the node that belongs to it.
+    pub(crate) account_to_index: BoundedBTreeMap<T::AccountId, u32, LiquidityTreeMaxNodes<U>>,
+    /// A vector that contains the indices of abandoned nodes. Sorted in the order in which the
+    /// nodes were abandoned, with the last element of the vector being the most recently abandoned
+    /// node.
+    pub(crate) abandoned_nodes: BoundedVec<u32, LiquidityTreeMaxNodes<U>>,
+}
+
+impl<T, U> LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    /// Create a new liquidity tree.
+    ///
+    /// # Parameters
+    ///
+    /// - `account`: The account to which the tree's root belongs.
+    /// - `stake`: The stake of the tree's root.
+    pub(crate) fn new(
+        account: T::AccountId,
+        stake: BalanceOf<T>,
```

Comment:
I agree. But the underlying "problem" here is that `T` is used as generic parameter. Once that's the case, I feel like using `BalanceOf<T>` is quite canonical. The reasons why I'm using `T` instead of generic types for account IDs, balance, etc.:

- _Maybe_ the tree will can to access some config trait in the future.
- The rust type inference has some serious issues with complex types. I've had big trouble when I tried to convert `Pool<T>` into `Pool<AccountId, Balance, ...>` and ended up litering the entire code with type hints before giving up and reverting to using the config. I see that other parachain projects use the same pattern.

Reply:
Using `T` also forces the user of the library to use it in the context of Substrate, which already is a big restriction. I think this is tolerable though, if it is designed to be used in a Substrate context anyways. The difference I wanted to point out is that while `T::AccountId` only requires `frame_system`, which is a fundamental building block of a Substrate project, `BalanceOf<T>` forces the developer to add a `MultiCurrency` implementation to the runtime, even other means of handling tokens are used. However, after all we just want a balance type, but not a complete implementation of the `MultiCurrency` trait.
Since usability in other projects was not a requirement, I don't request any changes. Just wanted to point that out.